libsolidity/interface/CompilerStack.cpp

```diff
@@ -607,6 +607,8 @@ bool CompilerStack::analyze()
 					modelChecker.analyze(*source->ast);
 			m_unhandledSMTLib2Queries += modelChecker.unhandledQueries();
 		}
+
+		m_stackState = AnalysisPerformed;
```

Comment:
For some additional context for why I think this change is correct:

This function can either return early above, in cases we can't guarantee even partial valid AST annotations - otherwise it tracks non-fatal errors in the ``noErrors`` boolean and returns a failure condition, while still setting the state to ``AnalysisPerformed``.

Fatal errors should be treated the same as the other conditions above, in which there's an early return without changing the stack state.

Reply:
But doing this bypasses the `m_stackState >= AnalysisPerformed` check at the beginning. It means that now you can run analysis the second time, over the dirty state that the previous run left.

I think it might be better to instead keep the current interpretation of `AnalysisPerformed` (i.e. literally performed but not necessarily successfully) and just replace these `return`s with throwing `FatalError` so that they hit the `catch` block too.

Reply:
The situation is similar with the `Parsed` state. It can be reached independently of whether there are errors or not: https://github.com/ethereum/solidity/blob/f1d2eda7953e56cb9456446898cdb5ff7628b1c2/libsolidity/interface/CompilerStack.cpp#L393-L398

Reply:
I.e. currently ``AnalysisPerformed`` is **not** set whenever you have already started analysis once anyways.

Reply:
I mean, we can also just split it into ``AnalysisPerformed``` to be set in the beginning of the function and ``AnalysisSuccessful`` in the end, if there's no errors...

Reply:
> I mean, we can also just split it into `AnalysisPerformed` to be set in the beginning of the function and `AnalysisSuccessful` in the end, if there's no errors...

Let me check how much of a hassle it is to do that...

Reply:
Well, it would need a case-by-case analysis of around 30 cases...

Reply:
git blame on ``AnalysisPerformed`` and you'll see that it's not exactly a stretch :-)